icu4c/source/common/uresbund.cpp

```diff
@@ -91,6 +91,172 @@ static UBool chopLocale(char *name) {
     return FALSE;
 }
 
+static UBool hasScript(const char* localeID) {
+    UErrorCode err = U_ZERO_ERROR;
+    int32_t scriptLength = uloc_getScript(localeID, NULL, 0, &err);
+    return scriptLength != 0;
+}
+
+static UBool hasRegion(const char* localeID) {
+    UErrorCode err = U_ZERO_ERROR;
+    int32_t scriptLength = uloc_getCountry(localeID, NULL, 0, &err);
+    return scriptLength != 0;
+}
+
+static UBool hasVariant(const char* localeID) {
+    UErrorCode err = U_ZERO_ERROR;
+    int32_t scriptLength = uloc_getVariant(localeID, NULL, 0, &err);
+    return scriptLength != 0;
+}
+
+// This file contains the tables for doing locale fallback, which are generated
+// by the CLDR-to-ICU process directly from the CLDR data.  This file should only
+// ever be included from here.
+#include "localeFallbackTables.h"
+
+static const char* performFallbackLookup(const char* key,
+                                         const char* keyStrs,
+                                         const char* valueStrs,
+                                         const int32_t* lookupTable,
+                                         int32_t lookupTableSize) {
+    const int32_t* bottom = lookupTable;
+    const int32_t* top = lookupTable + lookupTableSize;
+
+    while (bottom < top) {
+        const int32_t* middle = bottom + ((top - bottom) / 2);
+        middle -= (middle - lookupTable) % 2;
+        const char* entryKey = &(keyStrs[*middle]);
+        int32_t strcmpResult = uprv_strcmp(key, entryKey);
+        if (strcmpResult == 0) {
+            return &(valueStrs[middle[1]]);
+        } else if (strcmpResult < 0) {
+            top = middle;
+        } else {
+            bottom = middle + 2;
+        }
+    }
+    return nullptr;
+}
+
+static void removeScript(const char* localeID, char* localeIDWithoutScript) {
+    char workingLocaleID[ULOC_FULLNAME_CAPACITY];
+    char fieldBuffer[ULOC_FULLNAME_CAPACITY];
+    int32_t fieldLength = 0;
+    UErrorCode err = U_ZERO_ERROR;
+    
+    *localeIDWithoutScript = '\0';
+    uloc_getLanguage(localeID, fieldBuffer, ULOC_FULLNAME_CAPACITY, &err);
+    uprv_strcpy(workingLocaleID, fieldBuffer);
```

Comment:
Do we need to check err here? What if someone passes in a localeID that is longer than ULOC_FULLNAME_CAPACITY, or one that is malformed in some way? Is that checked at a higher level?

Reply:
This is a tough one.  All of my new code depends on the locale ID being well-formed, and I'm kind of relying on whatever existing mechanisms are present in the resource-bundle code to ensure that's true.  If we can be sure the existing resource-bundle code is giving us a well-formed locale ID, then what I have should be okay, since no external callers should be calling it.

But just in case, I went back and fixed this function to make sure the language and region codes are no longer than they're supposed to be, and to check the error codes coming back from both uloc_getLanguage() and uloc_getCountry().  If we get errors from either, this function will end up returning the original locale ID unchanged (i.e., it might still have a script code), which may cause additional problems, but at least we should be protected against writing off the end of the buffer.  Please take another look-- if you think there's a better way to do this, I'm all for it.